### DISCUSSION TITLE
SML script Honeywell AS2018 "-" Zeichen fehlt bei Grundlast (PV- Überschuss Einspeisung)

### DISCUSSION THREAD

**Discussion by OlafWalther1969** [arendst/Tasmota#19193] [Q&A]
Hello, I am using the Tasmota Smartmeter Firmware 12.3.1 with an IR reading head "Hichi Smartmeter" and a Honeywell AS2018 electricity meter.
All values ​​are displayed correctly, but not the base load with excess PV. The only thing missing here is the minus sign in front of the base load.

Example 1 Purchased electricity: Electricity meter shows 650W and Tasmota also shows 650W.
Example 2 Grid feed-in (PV surplus): Electricity meter shows -435W and Tasmota 435W without the minus in advance.

Is there a solution?

Here is the pasted script:

Hallo, ich habe die Tasmota Smartmeter Firmware 12.3.1 mit einem IR Lesekopf "Hichi Smartmeter" und einem Stromzähler Honeywell AS2018 im Einsatz.
Alle Werte werden richtig angezeigt, jedoch die Grundlast bei PV Überschuss nicht. Hier fehlt nur das Minuszeichen vor der Grundlast.

Bespiel 1 Netzbezug: Stromzähler zeigt 650W und Tasmota zeigt auch 650W an.
Beispiel 2 Netzeinspeisung (PV Überschuss): Stromzähler zeigt -435W und Tasmota 435W ohne das Minus vorweg.

Gibt es viellelicht eine Lösung?

Hier ist das eingefügte Script:

>D
scnt=0
res=0

>B
=>sensor53 r

>F
; count 100ms
scnt+=1
switch scnt

;300ms after start: set sml driver to 300 baud and send /?! as HEX to trigger the Meter
case 3
res=sml(1 0 300)
res=sml(1 1 "2F3F210D0A")

;1700ms later: Ack and ask for switching to 9600 baud
case 20
res=sml(1 1 "063035300D0A")

;300ms later: switch sml driver to 9600 baud
case 23
res=sml(1 0 9600)

;6000ms after start: Restart sequence
case 60
scnt=0

ends

>M 1
+1,3,o,0,9600,AS2018,1

1,0.9.1(@#),Zeitstempel,Uhr,timestamp,1
1,31.7.0(@1,Strom Phase 1,A,Current_L1,2
1,51.7.0(@1,Strom Phase 2,A,Current_L2,2
1,71.7.0(@1,Strom Phase 3,A,Current_L3,2
1,1.8.0(@1,Netzbezug gesamt,kWh,Netzbezug_gesamt,3
1,1.8.0*96(@1,Netzbezug Vortag,kWh,Netzbezug_1d,3
1,1.8.0*97(@1,Netzbezug 7Tage,kWh,Netzbezug_7d,3
1,1.8.0*98(@1,Netzbezug 30Tage,kWh,Netzbezug_30d,3
1,1.8.0*99(@1,Netzbezug 365Tage,kWh,Netzbezug_365d,3
1,2.8.0(@1,Netzeinspeisung gesamt,kWh,Netzeinspeisung_gesamt,3
1,2.8.0*96(@1,Netzeinspeisung Vortag,kWh,Netzeinspeisung_Vortag,3
1,2.8.0*97(@1,Netzeinspeisung 7Tage,kWh,Netzeinspeisung_7d,3
1,2.8.0*98(@1,Netzeinspeisung 30Tage,kWh,Netzeinspeisung_30d,3
1,2.8.0*99(@1,Netzeinspeisung 365Tage,kWh,Netzeinspeisung_365d,3
1,16.7.0(@1,aktuell P,kW,aktuell_P,3
1,96.1.0(@1,Zählernummer,,Meter_number,0
![tasmota_kein_Minus](https://github.com/arendst/Tasmota/assets/66950616/41e52902-c3dd-4252-9176-ef02727c1e1c)

**Comment by Jason2866** [comment]
@OlafWalther1969 This is an English only place. Anyways @gemu2015 Can you help?

**Reply by OlafWalther1969** [reply]
Sorry Jason!

I just modified the post.

**Comment by gemu2015** [comment]
you need to debug your raw meter data (see docs sensor53 d0) in condition import and export.
as this is asci obis, data is human readable and should be easily interpreted.
you will have to find an obis code that reflects import, export.
obviously 16.7.0 does not do it.

**Reply by OlafWalther1969** [reply]
Thanks for your answer.

16.7.0 shows the correct value when fed in, only the preceding "-" character is missing.
Is it because of the Obis protocol or because of Tasmota that no minus sign is transmitted?

best regards

**Reply by gemu2015** [reply]
please post the debug data for import and export case

**Reply by OlafWalther1969** [reply]
Hello @gemu2015,
thank you for your support!

here is a copy of the tasmota console (electricity purchase 16.7.0 "aktuell_P"2.5680W):

10:10:47.397 MQT: tele/S0IR/SENSOR = {"Time":"2023-07-27T10:10:47","AS2018":{"timestamp":"09:13:48","Current_L1":7.24,"Current_L2":0.95,"Current_L3":4.29,"Netzbezug_gesamt":4789.292,"Netzbezug_1d":19.134,"Netzbezug_7d":122.393,"Netzbezug_30d":570.601,"Netzbezug_365d":0.000,"Netzeinspeisung_gesamt":3223.950,"Netzeinspeisung_Vortag":22.066,"Netzeinspeisung_7d":99.362,"Netzeinspeisung_30d":687.698,"Netzeinspeisung_365d":0.000,"aktuell_P":2.580,"Meter_number":42048674,"power_in":4290,"power_in":0.0}}


here is a copy of the tasmota console (power feed 16.7.0 "aktuell_P"7.270W): (Here is the Minus - Symbol Missing)

13:57:07.666 MQT: tele/S0IR/SENSOR = {"Time":"2023-07-28T13:57:07","AS2018":{"timestamp":"13:00:07","Current_L1":11.97,"Current_L2":11.80,"Current_L3":9.49,"Netzbezug_gesamt":4823.406,"Netzbezug_1d":31.518,"Netzbezug_7d":130.149,"Netzbezug_30d":578.731,"Netzbezug_365d":0.000,"Netzeinspeisung_gesamt":3226.896,"Netzeinspeisung_Vortag":0.557,"Netzeinspeisung_7d":86.163,"Netzeinspeisung_30d":661.583,"Netzeinspeisung_365d":0.000,"aktuell_P":7.270,"Meter_number":42048674,"power_in":9490,"power_in":0.0}}

The "1,16.7.0(@1,aktuell P,kW,aktuell_P,3" shows the right valaue, but if the valaue is changing to negativ, the Minus Symbol is mising.


I´m was not able to do a schrennshot from the smartphone and the tasmaota consol at the same time. 



I´m not able to post raw data from the house smartmeter. I´m using an IR Hichi Smartmeter with Tasmota.

I know that 16.7.0 is the balance between electricity purchase and feed-in

**Reply by gemu2015** [reply]
sorry, that's not what i need.  Go to the tasmota console, type sensor53 d1 to log the raw sensor data, for both cases, import and export condition. post these logs.
after that return to normal operation with sensor53 d0

**Reply by OlafWalther1969** [reply]
Wow, i´ve don´t know the sensor53 d1 - 0 command. Thank You very much!
Here is a log during feed in:
14:38:36.875 CMD: sensor53 d1
14:38:36.882 MQT: stat/S0IR/RESULT = {"Time":"2023-07-28T14:38:36","SML":{"CMD":"dump: 1"}}
14:38:37.075 : 96.77.6*1(00-00-00,00 D|96.89.1(5530010236)
14:38:37.107 : 96.95(1261)
14:38:37.123 : 128.95(56593)
14:38:37.261 : 128.96(04869)
14:38:37.275 : 0.8.0(15*min)
14:38:37.293 : 0.9.1(13:41:42)
14:38:37.312 : 0.9.2(23-07-28)
14:38:37.322 : 0.9.5(5)
14:38:37.347 : 1.8.0(004823.406*kWh)
14:38:37.373 : 1.8.0*1(004304.899*kWh)
14:38:37.400 : 1.8.0*2(003730.955*kWh)
14:38:37.578 : 1.8.0*96(000031.518*kWh)
14:38:37.758 : 1.8.0*97(000130.149*kWh)
14:38:37.939 : 1.8.0*98(000578.731*kWh)
14:38:37.966 : 1.8.0*99(000000.000*kWh)
14:38:37.995 : 1.8.0*100(004823.406*kWh)
14:38:38.034 : 1.6.0(007.588*kW)(23-07-06,08:15)
14:38:38.073 : 1.6.0*1(007.112*kW)(23-06-15,07:30)
14:38:38.113 : 1.6.0*2(007.260*kW)(23-05-11,07:30)
14:38:38.137 : 2.8.0(003228.935*kWh)
14:38:38.163 : 2.8.0*1(002610.399*kWh)
14:38:38.190 : 2.8.0*2(001820.120*kWh)
14:38:38.229 : 2.6.0(007.580*kW)(23-07-02,15:45)
14:38:38.269 : 2.6.0*1(007.368*kW)(23-06-03,15:15)
14:38:38.309 : 2.6.0*2(007.912*kW)(23-05-02,14:15)
14:38:38.481 : 2.8.0*96(000000.557*kWh)
14:38:38.663 : 2.8.0*97(000086.163*kWh)
14:38:38.844 : 2.8.0*98(000661.583*kWh)
14:38:38.871 : 2.8.0*99(000000.000*kWh)
14:38:38.901 : 2.8.0*100(003228.937*kWh)
14:38:38.920 : 31.7.0(008.71*A)
14:38:38.941 : 51.7.0(008.44*A)
14:38:38.958 : 71.7.0(007.05*A)
14:38:38.975 : 32.7.0(240.0*V)
14:38:38.993 : 52.7.0(235.9*V)
14:38:39.012 : 72.7.0(238.9*V)
14:38:39.032 : 16.7.0(003.870*kW)
14:38:39.054 : 36.7.0(001.510*kW)
14:38:39.075 : 56.7.0(001.420*kW)
14:38:39.096 : 76.7.0(000.940*kW)
14:38:39.107 : 96.6.1(1)
14:38:39.127 : 96.1.3(21-09-09)
14:38:39.148 : 96.2.5(21-09-09)
14:38:39.172 : 96.2.2(21-09-09,06:43)
14:38:39.197 : 96.70(00-00-00,00:00)
14:38:39.227 : 96.71(00-00-00,00:00)(00)
14:38:39.257 : 96.71*1(00-00-00,00:00)(00)
14:38:39.290 : 96.71*2(00-00-00,00:00)(00)
14:38:39.302 : 96.7.0(01)
14:38:39.458 MQT: tele/S0IR/STATE = {"Time":"2023-07-28T14:38:39","Uptime":"0T00:03:08","UptimeSec":188,"Heap":19,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"FRITZ!Box 7490 Voehrum","BSSId":"A2:05:43:BA:69:81","Channel":6,"Mode":"11n","RSSI":52,"Signal":-74,"LinkCount":1,"Downtime":"0T00:00:03"}}
14:38:39.468 MQT: tele/S0IR/SENSOR = {"Time":"2023-07-28T14:38:39","AS2018":{"timestamp":"13:41:36","Current_L1":3.96,"Current_L2":3.65,"Current_L3":2.37,"Netzbezug_gesamt":4823.406,"Netzbezug_1d":31.518,"Netzbezug_7d":130.149,"Netzbezug_30d":578.731,"Netzbezug_365d":0.000,"Netzeinspeisung_gesamt":3228.932,"Netzeinspeisung_Vortag":0.557,"Netzeinspeisung_7d":86.163,"Netzeinspeisung_30d":661.583,"Netzeinspeisung_365d":0.000,"aktuell_P":2.190,"Meter_number":42048674}}
14:38:39.495 : 96.77.0*1(21-09-09,06:44,23-01-10,13:56)
14:38:39.621 : 96.77.0*2(00-00-00,00:00,00-00-00,00:00)
14:38:39.634 : 96.7.1(00)
14:38:39.678 : 96.77.1*1(00-00-00,00:00,00-00-00,00:00)
14:38:39.725 : 96.77.1*2(00-00-00,00:00,00-00-00,00:00)
14:38:39.739 : 96.7.2(00)
14:38:39.783 : 96.77.2*1(00-00-00,00:00,00-00-00,00:00)
14:38:39.830 : 96.77.2*2(00-00-00,00:00,00-00-00,00:00)
14:38:39.843 : 96.7.3(00)
14:38:39.889 : 96.77.3*1(00-00-00,00:00,00-00-00,00:00)
14:38:39.934 : 96.77.3*2(00-00-00,00:00,00-00-00,00:00)
14:38:39.949 : 96.77.4(00)
14:38:39.994 : 96.77.4*1(00-00-00,00:00,00-00-00,00:00)
14:38:40.040 : 96.77.4*2(00-00-00,00:00,00-00-00,00:00)
14:38:40.054 : 96.77.5(76)
14:38:40.101 : 96.77.5*1(23-07-28,11:49,99-99-99,99:99)
14:38:40.147 : 96.77.5*2(23-07-28,11:13,23-07-28,11:49)
14:38:40.161 : 96.77.6(00)
14:38:40.205 : 96.77.6*1(00-00-00,00:00,00-00-00,00:00)
14:38:47.388 : 
14:38:49.083 : 96.89.1(5530010236)
14:38:49.103 : 96.1.0(42048674)
14:38:49.118 : 96.95(1261)
14:38:49.133 : 128.95(56593)
14:38:49.274 : 128.96(04869)
14:38:49.291 : 0.8.0(15*min)
14:38:49.308 : 0.9.1(13:41:54)
14:38:49.327 : 0.9.2(23-07-28)
14:38:49.336 : 0.9.5(5)
14:38:49.361 : 1.8.0(004823.406*kWh)
14:38:49.389 : 1.8.0*1(004304.899*kWh)

**Reply by gemu2015** [reply]
As you can see 16.7.0 does not give a negative value!
check all the other obis codes and find one that indicates if the meter exports or imports power.
Or ask the manufacturer which value to use

**Reply by OlafWalther1969** [reply]
Thank you very much gemu2015,
i think the raw data helps me further.
Today, i have learned a lot.

best regards

**Comment by H3B0** [comment]
Hi Olaf, as I can see by the format of your Zählernummer you are fellow customer of the Stadtwerke Peine. And by the way, your log tells you live in Vöhrum. I just want to remind you that this is sensitive data. If it is egal to you, nevermind, sorry to bother you. Greetings from Dungelbeck

**Reply by OlafWalther1969** [reply]
Thank you for this information.
